File: NameSiloDnsUpdateService/EmailLoggingExtensions.cs
using Serilog;
using Serilog.Configuration;
using Serilog.Events;
using Serilog.Formatting;
using Serilog.Sinks.Email;
using System;
using System.Net;

namespace NameSiloDnsUpdateService
{
    public static class EmailLoggingExtensions
    {
        public static LoggerConfiguration Email(this LoggerSinkConfiguration loggerConfiguration, MyEmailConnectionInfo myConnectionInfo, string outputTemplate = "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level}] {Message}{NewLine}{Exception}", LogEventLevel restrictedToMinimumLevel = LogEventLevel.Verbose, int batchPostingLimit = 100, TimeSpan? period = null, IFormatProvider formatProvider = null, string mailSubject = "Log Email") =>
            loggerConfiguration.Email((EmailConnectionInfo)myConnectionInfo, outputTemplate, restrictedToMinimumLevel, batchPostingLimit, period, formatProvider, mailSubject);
        public static LoggerConfiguration Email(this LoggerSinkConfiguration loggerConfiguration, MyEmailConnectionInfo myConnectionInfo, ITextFormatter textFormatter, LogEventLevel restrictedToMinimumLevel = LogEventLevel.Verbose, int batchPostingLimit = 100, TimeSpan? period = null, string mailSubject = "Log Email") =>
            loggerConfiguration.Email((EmailConnectionInfo)myConnectionInfo, textFormatter, restrictedToMinimumLevel, batchPostingLimit, period, mailSubject);

        public class MyEmailConnectionInfo : EmailConnectionInfo
        {
            public new NetworkCredential NetworkCredentials
            {
                get => (NetworkCredential)base.NetworkCredentials;
                set => base.NetworkCredentials = value;
            }
        }
    }
}

File: NameSiloDynDns/EmailLoggingExtensions.fs
module NameSiloDynDns.EmailLoggingExtensions

open Serilog
open Serilog.Configuration
open Serilog.Events
open Serilog.Formatting
open Serilog.Sinks.Email
open System
open System.Net

type MyEmailConnectionInfo() =
    inherit EmailConnectionInfo()

    member _.NetworkCredentials
        with get () = base.NetworkCredentials :?> NetworkCredential
        and set (value: NetworkCredential) = base.NetworkCredentials <- value

type LoggerSinkConfiguration with
    member loggerConfiguration.Email
        (
            myConnectionInfo: MyEmailConnectionInfo,
            ?outputTemplate: string,
            ?restrictedToMinimumLevel: LogEventLevel,
            ?batchPostingLimit: int,
            ?period: TimeSpan,
            ?formatProvider: IFormatProvider,
            ?mailSubject: string
        ) =
        loggerConfiguration.Email(
            myConnectionInfo :> EmailConnectionInfo,
            defaultArg outputTemplate "{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level}] {Message}{NewLine}{Exception}",
            defaultArg restrictedToMinimumLevel LogEventLevel.Verbose,
            defaultArg batchPostingLimit 100,
            Option.toNullable<TimeSpan> period,
            Option.toObj<IFormatProvider> formatProvider,
            defaultArg mailSubject "Log Email"
        )

    member loggerConfiguration.Email
        (
            myConnectionInfo: MyEmailConnectionInfo,
            textFormatter: ITextFormatter,
            ?restrictedToMinimumLevel,
            ?batchPostingLimit,
            ?period,
            ?mailSubject
        ) =
        loggerConfiguration.Email(
            myConnectionInfo :> EmailConnectionInfo,
            textFormatter,
            defaultArg restrictedToMinimumLevel LogEventLevel.Verbose,
            defaultArg batchPostingLimit 100,
            Option.toNullable<TimeSpan> period,
            defaultArg mailSubject "Log Email"
        )
